network/src/network.rs

```diff
@@ -426,30 +427,43 @@ impl NetworkState {
     /// this method is intent to check observed addr by dial to self
     pub(crate) fn try_dial_observed_addrs(&self, p2p_control: &ServiceControl) {
         let mut pending_observed_addrs = self.pending_observed_addrs.write();
-        let public_addrs = { self.public_addrs.read().clone() };
-        for addr in pending_observed_addrs
-            .drain()
-            .chain(public_addrs.into_iter())
-        {
+        if pending_observed_addrs.is_empty() {
+            let addrs = self.public_addrs.write();
+            // random get addr
+            let index = rand::random::<usize>() % addrs.len();
+            let addr = addrs.iter().nth(index).unwrap();
             trace!("try dial observed addr: {:?}", addr);
             if let Err(err) = self.dial_inner(
                 p2p_control,
-                addr,
+                addr.clone(),
                 TargetProtocol::Single(SupportProtocols::Identify.protocol_id()),
                 true,
             ) {
-                debug!("try_dial_observed_addrs error {}", err);
+                trace!("try_dial_observed_addrs fail {} on public address", err)
+            }
+        } else {
+            let addr = { pending_observed_addrs.iter().next().unwrap().clone() };
+            trace!("try dial observed addr: {:?}", addr);
+            match self.dial_inner(
+                p2p_control,
+                addr.clone(),
+                TargetProtocol::Single(SupportProtocols::Identify.protocol_id()),
+                true,
+            ) {
+                Err(err) => {
+                    trace!("try_dial_observed_addrs fail {} on pending observed", err);
```

Comment:
should remove the addr from `pending_observed_addrs` when dial error?

Reply:
> should remove the addr from `pending_observed_addrs` when dial error?

I think this error is generally that the peerid already exists in the dial list, not tentacle dial queue full, so I think it’s not necessary to delete it here

Reply:
The latest commit deals with this problem in a more understandable way